### PR TITLE
Evaluation stage of JaxModel

### PR DESCRIPTION
**Pull request by VIGNESHinZONE**
1. Added support for evaluation and metrics
2. Added tests for basic functions

Its still in WIP

### CODE REVIEW

File: deepchem/models/jax_models/jax_model.py [hunk 1]
```diff
@@ -272,6 +276,219 @@ def fit_generator(self,
...
+    This is the private implementation of prediction.  Do not
+    call it directly.  Instead call one of the public prediction
+    methods.
+
+    Parameters
+    ----------
+    generator: generator
+      this should generate batches, each represented as a tuple of the form
+      (inputs, labels, weights).
+    transformers: list of dc.trans.Transformers
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:295]: `List[dc.trans.Transformers`. It's better to use the actual type rather than a description

File: deepchem/models/jax_models/jax_model.py [hunk 2]
```diff
@@ -272,6 +276,219 @@ def fit_generator(self,
...
+      Transformers that the input data has been transformed by.  The output
+      is passed through these transformers to undo the transformations.
+    uncertainty: bool
+      specifies whether this is being called as part of estimating uncertainty.
+      If True, it sets the training flag so that dropout will be enabled, and
+      returns the values of the uncertainty outputs.
+    other_output_types: list, optional
+      Provides a list of other output_types (strings) to predict from model.
+    Returns:
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:304]: I think this should be 
```
Returns
-------
```
by the numpy style guide

File: deepchem/models/jax_models/jax_model.py [hunk 3]
```diff
@@ -272,6 +276,219 @@ def fit_generator(self,
...
+      this should generate batches, each represented as a tuple of the form
+      (inputs, labels, weights).
+    transformers: list of dc.trans.Transformers
+      Transformers that the input data has been transformed by.  The output
+      is passed through these transformers to undo the transformations.
+    output_types: String or list of Strings
+      If specified, all outputs of this type will be retrieved
+      from the model. If output_types is specified, outputs must
+      be None.
+    Returns:
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:395]:
```
Returns
-------
```

File: deepchem/models/tests/test_jax_model.py [hunk 4]
```diff
@@ -94,5 +97,112 @@ def bce_loss(pred, tar, w):
...
       learning_rate=0.001,
       log_frequency=2)
-  results = j_m.fit(dataset, nb_epochs=50, deterministic=True)
-  assert results < 1.0
+  _ = j_m.fit(dataset, nb_epochs=25, deterministic=True)
+  scores = j_m.evaluate(dataset, [metric])
+  assert scores[metric.name] > 0.8
+
+
+@pytest.mark.jax
```
Comment (rbharath) [inline deepchem/models/tests/test_jax_model.py:105]: Should this test be marked slow?

File: deepchem/models/tests/test_jax_model.py [hunk 5]
```diff
@@ -94,5 +97,112 @@ def bce_loss(pred, tar, w):
...
+
+    def __call__(self, x: jnp.ndarray):
+      x = self._network(x)
+      return x, jax.nn.sigmoid(x)
+
+  def f(x):
+    net = Encoder(1)
+    return net(x)
+
+  # Model Initilisation
```
Comment (rbharath) [inline deepchem/models/tests/test_jax_model.py:129]: typo: Initialization (I'm not sure if initialisation is valid British/Indian spelling, but we've stuck to American spelling in the code so far)

File: deepchem/models/tests/test_jax_model.py [hunk 6]
```diff
@@ -94,5 +97,112 @@ def bce_loss(pred, tar, w):
...
+      log_frequency=2)
+  j_m.fit(dataset, nb_epochs=1000)
+  prediction = np.squeeze(j_m.predict_on_batch(X))
+  assert np.array_equal(y, np.round(prediction))
+  metric = dc.metrics.Metric(dc.metrics.roc_auc_score)
+  scores = j_m.evaluate(dataset, [metric])
+  assert scores[metric.name] > 0.9
+
+
+def test_fit_use_all_losses():
```
Comment (rbharath) [inline deepchem/models/tests/test_jax_model.py:160]: Should have @pytest.mark.jax/@pytest.mark.slow?

**Review by rbharath — COMMENTED** [review]
Just took a quick look through to familiarize myself with the code :)

File: deepchem/models/tests/test_jax_model.py (hunk 6 above)
Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:160]: sorry my bad, i missed this one. @pytest.mark.jax

**Comment by VIGNESHinZONE** [comment]
@rbharath could you take a look at this, I have finished it.

File: deepchem/models/jax_models/jax_model.py [hunk 7]
```diff
@@ -272,6 +281,270 @@ def fit_generator(self,
...
+    ----------
+    dataset: dc.data.Dataset
+      Dataset to make prediction on
+    masks: int
+      the number of dropout masks to average over
+    Returns
+    -------
+    for each output, a tuple (y_pred, y_std) where y_pred is the predicted
+    value of the output, and each element of y_std estimates the standard
+    deviation of the corresponding element of y_pred
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:490]: Add blank lines between these blocks.

File: deepchem/models/jax_models/jax_model.py [hunk 8]
```diff
@@ -285,23 +558,36 @@ def _set_trainable_params(self, params: hk.Params, opt_state: optax.OptState):
     self.params = params
     self.opt_state = opt_state
 
-  def _create_gradient_fn(self, loss, model, optimizer, loss_outputs):
+  def _create_eval_fn(self, forward_fn, params):
+    """
+    Calls the function to evaulate the model
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:563]: typo: evaluate

File: deepchem/models/tests/test_jax_model.py [hunk 9]
```diff
@@ -39,14 +40,15 @@ def rms_loss(pred, tar, w):
...
-      model,
+      model.apply,
       params,
       criterion,
       batch_size=256,
       learning_rate=0.001,
       log_frequency=2)
-  results = j_m.fit(dataset, deterministic=True)
-  assert results < 0.5
+  _ = j_m.fit(dataset, nb_epochs=25, deterministic=True)
```
Comment (ncfrey) [inline deepchem/models/tests/test_jax_model.py:49]: about how long does this training take?

File: deepchem/models/tests/test_jax_model.py [hunk 10]
```diff
@@ -87,12 +89,191 @@ def bce_loss(pred, tar, w):
...
+
+    def __init__(self, output_size: int = 1):
+      super().__init__()
+      self._network = hk.nets.MLP([512, 256, 128, output_size])
+
+    def __call__(self, x: jnp.ndarray):
+      x = self._network(x)
+      return x, jax.nn.sigmoid(x)
+
+  def f(x):
```
Comment (ncfrey) [inline deepchem/models/tests/test_jax_model.py:124]: Because this workflow is quite a bit different than other frameworks, it would be great to add some detailed documentation to `JaxModel` that describes this process.

File: deepchem/models/tests/test_jax_model.py [hunk 11]
```diff
@@ -87,12 +89,191 @@ def bce_loss(pred, tar, w):
...
+    net = Encoder(1)
+    return net(x)
+
+  # Model Initilisation
+  model = hk.transform(f)
```
Comment (ncfrey) [inline deepchem/models/tests/test_jax_model.py:129]: If every `JaxModel` requires the model to be wrapped as a function, does it make sense to allow the user to input an `hk.Module` (which will be familiar to torch and tensorflow users) and then automatically wrap it as a function and do the transform into a model?

File: deepchem/models/tests/test_jax_model.py [hunk 12]
```diff
@@ -87,12 +89,191 @@ def bce_loss(pred, tar, w):
...
+  modified_inputs = jnp.array(
+      [x.astype(np.float32) if x.dtype == np.float64 else x for x in inputs])
+  params = model.init(rng, modified_inputs)
+
+  # Loss Function
+  criterion = lambda pred, tar, w: jnp.mean(optax.sigmoid_binary_cross_entropy(pred[0], tar))
+
+  # JaxModel Working
+  j_m = JaxModel(
+      model.apply,
```
Comment (ncfrey) [inline deepchem/models/tests/test_jax_model.py:142]: Here it's a bit odd to pass an attribute as a parameter. Can we pass simply a model as described above and handle this internally in `JaxModel`?

File: deepchem/models/tests/test_jax_model.py (hunk 9 above)
Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:49]: They are quite fast as the datasets are small and apart from that I'm running it on `jit` .

File: deepchem/models/tests/test_jax_model.py (hunk 10 above)
Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:124]: Yes, all Haiku Models require this conversion and that is why i have included this line for later PR
https://github.com/VIGNESHinZONE/deepchem/blob/cee0531ddd00e4ca541e45c26cd087bb0bbbc277/deepchem/models/jax_models/jax_model.py#L660

File: deepchem/models/tests/test_jax_model.py (hunk 12 above)
Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:142]: jax's core `grad` feature requires a pure function and the gradients are taken with respect to the parameters.

`model.apply` here is the pure forward function

Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:142]:
```
forward_fn, initilisation = hk.transform(f)
params = initilisation(rng, modified_inputs)
```
this here is the more descriptive way of writing it

File: deepchem/models/tests/test_jax_model.py (hunk 11 above)
Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:129]: Yes, I can add a utility function that does it internally.

File: deepchem/models/tests/test_jax_model.py (hunk 4 above)
Comment (VIGNESHinZONE) [reply deepchem/models/tests/test_jax_model.py:104]: It runs quite fast on CPU as well.

**Review by rbharath — COMMENTED** [review]
Did a quick pass. Seconding @ncfrey's comments that we'll need some detailed docs since the style here is quite different from our other models.

@peastman If you have bandwidth, your review would be very helpful here as well for selecting a right API for `JaxModel`

**Comment by VIGNESHinZONE** [comment]
Thanks for the review @ncfrey @rbharath , I will try adding docs explaining the different style. Seeing the reviews, I will include Two other test suites to further clarify the usages
1) Model with pure Jax
2) Model with sequential Jax

**Comment by VIGNESHinZONE** [comment]
@rbharath @ncfrey Could you take a look at this - 

1. I have commented out the uncertainty prediction
2. I have also modularised few parts of the code as well.

File: deepchem/models/jax_models/jax_model.py [hunk 13]
```diff
@@ -24,13 +28,57 @@
 logger = logging.getLogger(__name__)
 
 
+def create_default_eval_fn(forward_fn, params):
+  """
+    Calls the function to evaulate the model
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:33]: Spacing on the docstring looks off here

File: deepchem/models/jax_models/jax_model.py [hunk 14]
```diff
@@ -24,13 +28,57 @@
...
+    predict = forward_fn(params, rng, batch)
+
+    return predict
+
+  return eval_model
+
+
+def create_default_update_fn(optimizer, model_loss):
+  """
+    This function calls the update function, to implement the backpropogation
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:47]: Same here

File: deepchem/models/jax_models/jax_model.py [hunk 15]
```diff
@@ -24,13 +28,57 @@
...
+    updates, opt_state = optimizer.update(grads, opt_state)
+    new_params = optax.apply_updates(params, updates)
+    return new_params, opt_state, batch_loss
+
+  return update
+
+
+def create_default_gradient_fn(forward_fn, loss_outputs, loss_fn):
+  """
+    This function calls the gradient function, to implement the backpropogation
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:64]: Spacing

File: deepchem/models/jax_models/jax_model.py (hunk 13 above)
Comment (rbharath) [reply deepchem/models/jax_models/jax_model.py:33]: Typo: evaluate

File: deepchem/models/jax_models/jax_model.py (hunk 14 above)
Comment (rbharath) [reply deepchem/models/jax_models/jax_model.py:47]: Typo: backpropagation

File: deepchem/models/jax_models/jax_model.py [hunk 16]
```diff
@@ -40,24 +88,26 @@ class JaxModel(Model):
...
+               eval_fn: Callable = create_default_eval_fn,
+               rng=jax.random.PRNGKey(1),
                log_frequency: int = 100,
                **kwargs):
     """
     Create a new JaxModel
-
     Parameters
     ----------
     model: hk.State or Function
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:113]: Since we've removed `model`, we should remove/rename the docstring as well

File: deepchem/models/jax_models/jax_model.py [hunk 17]
```diff
@@ -272,6 +323,260 @@ def fit_generator(self,
...
 
+  def _predict(
+      self, generator: Iterable[Tuple[Any, Any, Any]],
+      transformers: List[Transformer], uncertainty: bool,
+      other_output_types: Optional[OneOrMany[str]]) -> OneOrMany[np.ndarray]:
+    """
+    Predict outputs for data provided by a generator.
+    This is the private implementation of prediction.  Do not
+    call it directly.  Instead call one of the public prediction
+    methods.
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:334]: Need a space here before the parameters section

File: deepchem/models/jax_models/jax_model.py [hunk 18]
```diff
@@ -272,6 +323,260 @@ def fit_generator(self,
...
+      (inputs, labels, weights).
+    transformers: List[dc.trans.Transformers]
+      Transformers that the input data has been transformed by.  The output
+      is passed through these transformers to undo the transformations.
+    uncertainty: bool
+      specifies whether this is being called as part of estimating uncertainty.
+      If True, it sets the training flag so that dropout will be enabled, and
+      returns the values of the uncertainty outputs.
+    other_output_types: list, optional
+      Provides a list of other output_types (strings) to predict from model.
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:348]: Need a space before the returns section

File: deepchem/models/jax_models/jax_model.py [hunk 19]
```diff
@@ -272,6 +323,260 @@ def fit_generator(self,
...
+    Returns
+    -------
+      a NumPy array of the model produces a single output, or a list of arrays
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:351]: Spacing here. 

Typo: "A NumpyArray if the model produces a single output, or a list of arrays otherwise."

File: deepchem/models/jax_models/jax_model.py [hunk 20]
```diff
@@ -272,6 +323,260 @@ def fit_generator(self,
...
+      if it produces multiple outputs
+    """
+    return self._predict(generator, transformers, False, output_types)
+
+  def predict_on_batch(self, X: ArrayLike, transformers: List[Transformer] = []
+                      ) -> OneOrMany[np.ndarray]:
+    """Generates predictions for input samples, processing samples in a batch.
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:451]: Need a space before parameters and before return sections

File: deepchem/models/jax_models/jax_model.py [hunk 21]
```diff
@@ -272,6 +323,260 @@ def fit_generator(self,
...
+
+    pass
+
+  def predict(
+      self,
+      dataset: Dataset,
+      transformers: List[Transformer] = [],
+      output_types: Optional[List[str]] = None) -> OneOrMany[np.ndarray]:
+    """
+    Uses self to make predictions on provided Dataset object.
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:478]: Spacing before parameters and returns here as well

**Review by rbharath — CHANGES_REQUESTED** [review]
Looks pretty good! Most comments are on docstrings and formatting
